data/account/participationRegistry.go

```diff
@@ -917,31 +932,52 @@ func (db *participationDB) Record(account basics.Address, round basics.Round, pa
 // Waiting for all asynchronous IO to complete includes actions from other threads.
 // Flush waits for the participation registry to be idle.
 // Flush returns the latest error generated by async IO, if any.
-func (db *participationDB) Flush() error {
-	db.mutex.Lock()
-	defer db.mutex.Unlock()
+func (db *participationDB) Flush(timeout time.Duration) error {
+	var err error
+	done := make(chan struct{})
+	go func() {
+		db.mutex.Lock()
+		defer db.mutex.Unlock()
 
-	db.flushesPending++
-	db.writeQueue <- partDBWriteRecord{
-		flush: true,
-	}
+		db.flushesPending++
+		db.writeQueue <- partDBWriteRecord{
```

Comment:
the writing of `db.writeQueue <- partDBWriteRecord` which is done under mutex lock, could deadlock with the flush thread trying to take the lock (`db.mutex.Lock()`). In fact, if you'll try to Flush very quickly many time, it's very likely to happen.

Reply:
Great catch, I think there were a few opportunities for this to happen with other functions as well.

I reproduced one of the deadlocks and added a separate mutex for the specific variable used by the flush thread.